[WIP] Terraform Docs for TF policy (#2690)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/reference/policy/policy.mdx

---
page_title: `policy`
description: >-
  Reference for the `policy {}` block used to configure Terraform policy evaluation settings.
---

# `policy`

The `policy` block configures how Terraform policy evaluates your policies, including which version of Terraform to use to perform the evaluation.

@include "partials/beta.mdx"

## Configuration model

```hcl
policy {
  terraform_config {
    required_version = "<version_constraint>"
  }
  plugins {
    <plugin_name> = {
      source = "<path_to_binary>"
    }
  }
}

> ```

Reviewer: ```suggestion
The `policy` block supports the following configuration:

- `policy` &nbsp block 
    - [`terraform_config`](#terraform_config) &nbsp block
        - [`required_version`](#required_version) &nbsp string  
    -  [`plugins`](#plugins) &nbsp block   
        - <plugin_name> &nbsp map  
            - `source` &nbsp string | required
```
The model is like a TOC of the keywords and arguments in the configuration. What you have for the existing model is what we would normally put into the complete configuration.  Have a look at one of the other language blocks as a reference: https://developer.hashicorp.com/terraform/language/block/action

## Complete configuration example

```hcl
policy {
  terraform_config {
    required_version = ">= 1.16.0"
  }
  plugins {
    example = {
      source = "../plugins/bin/example"
    }
    custom_validator = {
      source = "../plugins/bin/custom-validator"
    }
  }
}
```

## Specification

> ### `terraform_config` block

Reviewer: ```suggestion

A `policy` block supports the following configuration. 

### `terraform_config`
```

> The `terraform_config` block defines settings which are specific to Terraform.

Reviewer: ```suggestion
The `terraform_config` block defines settings specific to the Terraform CLI provisioning the infrastructure. The `terraform_config` block takes the [`required_version`](#required_version) as an argument.  

- Data type: Block
- Default: None
```
The principle we want to follow is that each line of configuration has its own explanation. The reference for the `terraform` block in the language reference splits these, too: https://developer.hashicorp.com/terraform/language/block/terraform

#### Arguments

| Argument | Required | Type | Description |
|----------|----------|------|-------------|

> | `required_version` | No | String | A version constraint that validates the current Terraform CLI version. Uses the same version constraint syntax as Terraform itself. |

Reviewer: ```suggestion
```

> The `required_version` attribute uses the same version constraint syntax as Terraform itself. Refer to [Version constraints - Version constraint syntax](/terraform/language/expressions/version-constraints#version-constraint-syntax) for more information.

Reviewer: ```suggestion

### `required_version`

The `required_version` attribute declares the Terraform version required to provision infrastructure. Refer to [Version constraint syntax](/terraform/language/expressions/version-constraints#version-constraint-syntax) for more information.

- Data type: String
- Default: None
- Example: [Require minimum Terraform version](#require-minimum-terraform-version) 
```
Or is the default `latest`?

### `plugins` block

The `plugins` block allows you to refer to binary plugins that provide custom functions you can use in your policies. Each plugin is defined as a named block within the `plugins` block.

#### Arguments

| Argument | Required | Type | Description |
|----------|----------|------|-------------|
| `source` | Yes | String | The relative path to a binary in the same VCS repository as your policies. The binary must be a valid Terraform policy Go plugin server. |

Refer to [Policy plugins](/terraform/policy/write/plugins) for more information about creating and using policy plugins.

In your policies, reference your functions with the syntax `plugin::<plugin_name>::<function_name>(<function_arguments>...)`.

## Examples

> 

Reviewer: ```suggestion
### `plugins`

Specifies plugins that provide custom functions to use in your policies. Plugins must be compiled binaries. Refer to [Policy plugins](/terraform/policy/write/plugins) for more information. 

Specify each plugin as a named block within the `plugins` block.

- Data type: Block
- Default: None
- Example: [Configure multiple plugins](#configure-multiple-plugins)

### `source`

Specifies the relative path to a plugin binary in the same VCS repository as your policies. The binary must be a valid Terraform policy Go plugin server. Refer to [Policy plugins](/terraform/policy/write/plugins) for more information about creating and using policy plugins.

In your policies, reference your functions with the syntax `plugin::<plugin_name>::<function_name>(<function_arguments>...)`.

- Data type: String
- Default: None
- This argument is required to use the plugin.
- Example: [Configure multiple plugins](#configure-multiple-plugins)

## Examples

The following examples show how to write policy configurations for common use cases.

```
